pkg/controller/persistentvolume/persistentvolume_framework_test.go

```diff
@@ -0,0 +1,462 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package persistentvolume
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+	"sync"
+	"testing"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/resource"
+	"k8s.io/kubernetes/pkg/api/testapi"
+	"k8s.io/kubernetes/pkg/client/cache"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/fake"
+	"k8s.io/kubernetes/pkg/client/testing/core"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/conversion"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/types"
+	"k8s.io/kubernetes/pkg/util/diff"
+)
+
+// This is a unit test framework for persistent volume controller.
+// It fills the controller with test claims/volumes and can simulate these
+// scenarios:
+// 1) Call syncClaim/syncVolume once.
+// 2) Call syncClaim/syncVolume several times (both simulating "claim/volume
+//    modified" events and periodic sync), until the controller settles down and
+//    does not modify anything.
+// 3) Simulate almost real API server/etcd and call add/update/delete
+//    volume/claim.
+// In all these scenarios, when the test finishes, the framework can compare
+// resulting claims/volumes with list of expected claims/volumes and report
+// differences.
+
+// controllerTest contains a single controller test input.
+// Each test has initial set of volumes and claims that are filled into the
+// controller before the test starts. The test then contains a reference to
+// function to call as the actual test. Available functions are:
+//   - testSyncClaim - calls syncClaim on the first claim in initialClaims.
+//   - testSyncClaimError - calls syncClaim on the first claim in initialClaims
+//                          and expects an error to be returned.
+//   - testSyncVolume - calls syncVolume on the first volume in initialVolumes.
+//   - any custom function for specialized tests.
+// The test then contains list of volumes/claims that are expected at the end
+// of the test.
+type controllerTest struct {
+	// Name of the test, for logging
+	name string
+	// Initial content of controller volume cache.
+	initialVolumes []*api.PersistentVolume
+	// Expected content of controller volume cache at the end of the test.
+	expectedVolumes []*api.PersistentVolume
+	// Initial content of controller claim cache.
+	initialClaims []*api.PersistentVolumeClaim
+	// Expected content of controller claim cache at the end of the test.
+	expectedClaims []*api.PersistentVolumeClaim
+	// Function to call as the test.
+	test testCall
+}
+
+type testCall func(ctrl *PersistentVolumeController, reactor *volumeReactor, test controllerTest) error
+
+const testNamespace = "default"
+const versionConflictError = "VersionError"
```

Comment:
Use errors.New() to define an error constant, *don't* define a constant and fmt.Errorf it.